clients/src/main/java/org/apache/kafka/clients/KafkaFuture.java

```diff
@@ -0,0 +1,321 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.clients;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * A flexible future which supports call chaining and other asynchronous programming patterns.
+ * This will eventually become a thin shim on top of Java 8's CompletableFuture.
```

Comment:
One use-case I can think of is cancelling a future.  In `CompletableFuture` this is essentially equivalent to a call to `CompletableFuture#completeExceptionally(new CancellationException(...))`  Cancellation isn't an optional operation for us since it's part of Java's original `java.util.concurrent.Future` API, which we need to implement.

I agree that it's a bit nicer conceptually to separate the "listen for stuff happening" API from the "signal that stuff has happened" API.  It makes sense that Scala did this with `scala.concurrent.Future`.  Java's `CompletableFuture` muddies the waters a bit by combining them into one concrete class.  I don't know if this is something that is really likely to trip users up, though...